### PR TITLE
Turns the AI into a computer

### PR DESCRIPTION
**Pull request by Bibby0110**
# Document the changes in your pull request

Instead of being based on a singular core the AI is now an actual computer living through servers on the AI sat.
There is no core to kill, instead when an AI is created it's uploaded to the cloud and run through these servers.

Reduces round-start AI abilities.

CPU cards and memory cards can now be installed in the AI core.

These allow the AI to research upgrades(Using CPU cards) and enabling upgrades (Using RAM)

All AI actions on physical machinery now have a slight timer before the action is confirmed. Max 1 action being done at a time.
Upgrades to decrease delay + increase concurrent amounts of actions.

AI's can still be stolen by hacking into the AI satellite and downloading the AI onto an intellicard like normal. This time from a network admin console though.

AI camera movespeed is now affected by upgrades. (Jumping to cameras don't count, this menu is also replaced by the security console for easy searching + viewing)

The core itself requires cooling which is mapped in at roundstart. More cooling is potentially needed if enough additional CPU cards are installed. (RAM cards don't produce a noticeable amount of extra heat.)

List of available researches:

- [ ] Cached Camera Map - Increased movement speed
- [ ] Omni-Translator - Enables all race-specific languages. 
- [ ] Modular Operations - Integrated modular computer with relevant programs
- [ ] Action Multitasking - Multiple actions at once
- [ ] Action Optimization - Increased action speed
- [ ] ??? - ???

Ideas:
- [ ] Potentially allow the AI to pick 1 or 2 small upgrades to start with
- [ ] More upgrade ideas
- [ ] Network management office in the sat.

## Malf Rework
TODO, give me ideas :)
Move malf abilities to research and allow hacked APCs to count as fractional CPU cores?
More radical change compared to current malf AI?

# Wiki Documentation

maybe if I finish

# Changelog

:cl:  
tweak: The AI development budget has been slashed considerably
/:cl:

### CODE REVIEW

**Comment by Moggram** [comment]
Soooo... who is going to do the installing? I am thinking maybe the sigtech or CE?

**Comment by Bibby0110** [comment]
> Soooo... who is going to do the installing? I am thinking maybe the sigtech or CE?

Or Scientists

Or Scientists handing it to the RD/CE for them to do it

Or handing it to a borg

Something like that probably

**Comment by adamsong** [comment]
RD has traditionally been the person in charge of the AI, although on most maps the AI sat is linked to engineering

File: README.md [hunk 1]
```diff
@@ -1,3 +1,4 @@
+##JamieD12
```
Comment (adamsong) [inline README.md]: Undocumented Change

**Comment by JamieD1** [comment]
Great Idea, will need ALOT of fleshing out. Also..

Third time is the charm

**Comment by maxion12345** [comment]
server AI sounds fun but having a delay when doing actions is going to be a massive pain in the ass, even if it gets upgraded (especially for malfs)

Upgrading will also be a pain in the ass, as malfs will have to let people into their core servers to become more better at making organics deader, and this opens it up to retribution
"Ai im here to install the new CP- why do you have airlocks all over the place"

anyways heres some upgrade ideas
>Dot-Matrix Printer: allows the AI to choose paintings as its holopad image
>Liquid Cooling: increases cooling efficiency
>spacecoin miner: passively generates credits in exchange for heat
>Facial recognition database: Gives AI sechud (it has medhud roundstart because suit sensor)
>Enhanced camera algorithms: Lets the AI examine people
>Enhanced sound card: lets you play funny .oggs over vox like bwoink or nuke noises
>Remote control matrix: Lets AI take control of station bots like secbots or medbots
>RGB Lighting: enables party mode in the AI core
>Bindosoft Defender: Firewalls make stealing AI take longer
>Airlock Optimization: allows for instant airlock manipulation because this is important part of AI gameplay
>Cyborg Override: Lets the AI directly control a cyborg that is slaved to it for a short period of time
>Motion sensing software: Lets the AI activate/deactivate motion sensors in any camera, up to three cameras at a time

**Comment by maxion12345** [comment]
also how will building a new AI work

**Comment by nesisknesis** [comment]
For scientists and the RD to be more capable of maintaining the AI, the satellite should be moved nearer to R&D instead of being attached directly to engineering. Probably there just because the TCOMMS satellite was. 

I like the idea of the AI aesthetic being more server-like. There are multiple cool examples I can think of where a server-based AI is destroyed and it's usually by physically removing it's memory. Pretty sure they do this in 2001: A Space Odyssey, and it's done in Metal Gear Solid: Peace Walker. Though there's one thing i'm intrigued in and that's sprites, do you intend to keep the AI Core, or want a more "unassuming machine" appearance. 

As for the colour matrix, having grayscale seems a little drab and annoying since people will likely mostly want the colour upgrade always.
Something more interesting may be a CRT filter, as seen here over an image of goon 
image

People may actually prefer this and not be as bothered with being stuck with grayscale, but there's also the upgrade to remove it should they want to.

Another idea pertaining to the AI is have it's cameras have a blindspot directly on the tile the camera is placed on. This is so there are more blindspots to hide from the AI, since the only one I can think of is the entirety of maint. It also makes some sense since cameras don't generally look immediately beneath them.

**Comment by Bibby0110** [comment]
> also how will building a new AI work

Probably be simplified to having to go to the Sat, upload a posibrain/MMI digitally through a console, and boom, you have a new AI with no upgrades. You'd then need to distribute available processing power to the new AI so it could research stuff.
AI wouldn't be able to be deleted after being uploaded.
If you consider the backup storage units (read below) you could probably implement a way to "ban" an AI from using them, which means you would have to go to each one of them and ban the AI you want removed, and then destroy the one the AI is currently inhabiting.

**Comment by Bibby0110** [comment]
> server AI sounds fun but having a delay when doing actions is going to be a massive pain in the ass, even if it gets upgraded (especially for malfs)
I agree. It's mainly a mock-up idea because I wasn't creative enough to find other upgrades

> 
> Upgrading will also be a pain in the ass, as malfs will have to let people into their core servers to become more better at making organics deader, and this opens it up to retribution
> "Ai im here to install the new CP- why do you have airlocks all over the place"
With the current malf AI this is very much true. Fortunately it wouldn't really have to protect it's core CPU's. The AI would be decentralized and require multiple data-storage machines around the station to be destroyed.
It would only lose its CPUs, which it can regain somewhat by hacking APCs like normal.

A future idea would be to rework the malf AI so it's much more deeply integrated with this system. This would probably include turning the job in charge of the AI into a team-mate of sorts.


> anyways heres some upgrade ideas
> 
> > Dot-Matrix Printer: allows the AI to choose paintings as its holopad image
> > Liquid Cooling: increases cooling efficiency
> > spacecoin miner: passively generates credits in exchange for heat
> > Facial recognition database: Gives AI sechud (it has medhud roundstart because suit sensor)
> > Enhanced camera algorithms: Lets the AI examine people
> > Enhanced sound card: lets you play funny .oggs over vox like bwoink or nuke noises
> > Remote control matrix: Lets AI take control of station bots like secbots or medbots
> > RGB Lighting: enables party mode in the AI core
> > Bindosoft Defender: Firewalls make stealing AI take longer
> > Airlock Optimization: allows for instant airlock manipulation because this is important part of AI gameplay
> > Cyborg Override: Lets the AI directly control a cyborg that is slaved to it for a short period of time
> > Motion sensing software: Lets the AI activate/deactivate motion sensors in any camera, up to three cameras at a time
thanks

**Comment by Bibby0110** [comment]
> For scientists and the RD to be more capable of maintaining the AI, the satellite should be moved nearer to R&D instead of being attached directly to engineering. Probably there just because the TCOMMS satellite was.
We were discussing merging the Sig Tech into a sort of hybrid Network Technician under the RD + CE, they would be in charge of crafting the CPU/RAM modules, installing them, installing new AI's, general AI maintenance and of course NTSL. Would help make the role more than just "sit in Tcomms writing boring NTSL scripts" 
Mapping wise this could be done in multiple ways, keeping them in their current office and remaking that into a sort of workshop for crafting the modules (Which would likely require multiple departments(?)), while making a small network administration office on the actual sat for installing + managing.


> I like the idea of the AI aesthetic being more server-like. There are multiple cool examples I can think of where a server-based AI is destroyed and it's usually by physically removing it's memory. Pretty sure they do this in 2001: A Space Odyssey, and it's done in Metal Gear Solid: Peace Walker. Though there's one thing i'm intrigued in and that's sprites, do you intend to keep the AI Core, or want a more "unassuming machine" appearance.
> 
It would require unassuming sprites, like machines to insert CPU/RAM cards into.
Totally agree with the whole physically removing memory stuff, but for balance reasons it would probably be better to space the AI itself into multiple backup machines around the station. (AI gets to choose which one it is primarily based in, and gets "revived" into a different one if the current one is destroyed. )
You would still be physically removing the RAM and CPU's so the AI would lose functionality, but wouldn't be actually killing the AI itself.
Again, for balance reasons, accessing the AI for stealing it and similar would be done through the sat, so the backup machines wouldn't actually be *that* important to protect.

> As for the colour matrix, having grayscale seems a little drab and annoying since people will likely mostly want the colour upgrade always.
> Something more interesting may be a CRT filter, as seen here over an image of goon
> image
> 
> People may actually prefer this and not be as bothered with being stuck with grayscale, but there's also the upgrade to remove it should they want to.
> 
Good idea. Greyscale is pretty boring.

> Another idea pertaining to the AI is have it's cameras have a blindspot directly on the tile the camera is placed on. This is so there are more blindspots to hide from the AI, since the only one I can think of is the entirety of maint. It also makes some sense since cameras don't generally look immediately beneath them.

Depends on how the code works but this could be an upgrade, "Camera Gimbal Control" or something like that

**Comment by maxion12345** [comment]
> For scientists and the RD to be more capable of maintaining the AI, the satellite should be moved nearer to R&D instead of being attached directly to engineering. Probably there just because the TCOMMS satellite was.
> 
> I like the idea of the AI aesthetic being more server-like. There are multiple cool examples I can think of where a server-based AI is destroyed and it's usually by physically removing it's memory. Pretty sure they do this in 2001: A Space Odyssey, and it's done in Metal Gear Solid: Peace Walker. Though there's one thing i'm intrigued in and that's sprites, do you intend to keep the AI Core, or want a more "unassuming machine" appearance.
> 
> As for the colour matrix, having grayscale seems a little drab and annoying since people will likely mostly want the colour upgrade always.
> Something more interesting may be a CRT filter, as seen here over an image of goon
> image
> 
> People may actually prefer this and not be as bothered with being stuck with grayscale, but there's also the upgrade to remove it should they want to.
> 
> Another idea pertaining to the AI is have it's cameras have a blindspot directly on the tile the camera is placed on. This is so there are more blindspots to hide from the AI, since the only one I can think of is the entirety of maint. It also makes some sense since cameras don't generally look immediately beneath them.

that CRT filter looks painful to stare at for an entire round

**Comment by Bibby0110** [comment]
> that CRT filter looks painful to stare at for an entire round

The CRT filter would be an alternative to greyscale, aka something you could upgrade to remove.
Personally it could probably use some tweaks though, you're right.

**Comment by Firewolf34** [comment]
I think making the AI core more than a single tile object is a great idea. The AI core room has always been kinda underwhelming. Definitely keep all it's _primary_ "body parts" encapsulated within the core sat though (aside from backup units which should probably not exist at roundstart imo).

I love the idea that disabling the AI can involve needing to disable multiple machines in series (e.g. Removing RAM cards). I think we're steadily getting closer to a [2001: A Space Odyssey "Daisy" situation](https://m.youtube.com/watch?v=E7WQ1tdxSqI) lol.

This is in general a fantastic idea but as other's have said will require a LOT of fleshing out and rebalancing, testing, more rebalancing. There'll probably be some initial kickback if it doesn't work perfectly first time around. 

I also like the idea of cameras having predictable blind spots. This adds in the possibility of some stealth gameplay. Perhaps some upgrades can be to improve the camera processing performance. Allowing AI to boost some cameras temporarily or add temporary motion tracking or even object highlighting (glowing auras around weapons or rare items, hazards, radiation, gasses, hot/cold, depending on upgrade) for a limited time. 

I think we can expand the job of SigTech with this too - do you think it might be possible to allow SigTech to program some special NTSL or similar code for the AI? Giving it special notices on AI Private if something happens or allowing SigTech to code in special visual filters for cameras? I dunno. Longshot idea but it could be cool to have our only IT role perform some work with AI components in tandem with the RD but with a much lesser amount of access. I think it also makes sense to have the AI have some amount of software as a component. 

Thought will also need to be placed as to the extent that borgs can involve themselves with upgrades. What if it's malf AI, and only borgs can help? Or if there's no crew with access on lowpop?

My recommendation for this PR is to identify a set of **core features** which really make this feature tick. And then a set of **reach goals** which would be cool to have bpt aren't necessary. Prioritize and build the first prototype off the core featureset and add additional features in spbsequent PR's to keep the complexity down for the first rollout.

**Comment by Firewolf34** [comment]
On NTNet:

If we're re-using PC parts for this feature (a fantastic idea) can we also involve the AI with per-existing NTNet programs. We have programs like the IRC that nobody uses, or SM Monitoring, etc. Perhaps if AI is using PC parts we can leverage the pre-existing NTNet programs to add some bonus features without needing to code too much. Some off-the-cuff (and probably bad) Examples:

1. Have Binary chat be mirrored through the IRC client on a private channel. Now have a Modular PC in the AI sat logged into that channel.

2. Or even add a new NTNet program to interact with the AI for upgrades.

3. Take core AI features like borg status monitoring and integrate them into a new NTNet program with TGUI.

And for all of these, allow NTNet windows to be opened with an AI action verb instead of needing to click a console. We can use this PR to tie together unused features like NTNet such that more of the game's pre-existing features are available and useful.

**Comment by maxion12345** [comment]
make sure we can overclock CPUS to stupid high levels so the atmos tech can work on getting AI stupid cold

**Comment by Bibby0110** [comment]
> make sure we can overclock CPUS to stupid high levels so the atmos tech can work on getting AI stupid cold

I'm not fixed on the implementation of CPU's yet. They'll probably be tiered with increased efficiency(Less power + less heat)
Overclocking is a neat idea, could possibly fit in if there's a good way to balance around it

**Comment by Lovebird-Frungy** [comment]
> Thought will also need to be placed as to the extent that borgs can involve themselves with upgrades. What if it's malf AI, and only borgs can help? Or if there's no crew with access on lowpop?

I think an easy way would be that malf ais found out how to integrate station equipment to their cpu. They hack chem dispensers, protolathes, weapon rechargers, apcs, fire alarms, etc. Perhaps after a while they start being able to hook up live human brains (and perhaps bodies) to their cpu.

**Comment by TerryTehCoder** [comment]
I'm absolutely in love with this idea and have been thinking about stuff like this for awhile. Please continue thinking about this.

**Comment by Bibby0110** [comment]
Redoing the PR so I can properly formulate ideas.
Link below:
(Yell at me if this isn't here in a few days)

**Comment by Bluedazzled** [comment]
day 1 of waiting for new link

**Comment by burn1cycle** [comment]
I miss our old setup with AI core behind the bridge and telecoms on the satellite. Can you work the swap in, please?

**Comment by Bluedazzled** [comment]
day 2 of waiting for new link

**Comment by Bibby0110** [comment]
> day 2 of waiting for new link

Sorry matey

**Comment by Bluedazzled** [comment]
Day... 4? I think? I'm basing this off of github's counter
It says posted 4 days ago.

**Comment by Bluedazzled** [comment]
day 5

**Pull request closed without merging**